### ISSUE TITLE
How can I remove entry from zip?

### ISSUE THREAD

**Issue by RocksonZeta** [kuba--/zip#14]
see the title!

**Comment by kuba--** [comment]
Hey, @RocksonZeta - are you trying to remove some entry from existing archive? So far, there is no API which let you do it.
You can find only reads and writes and thanks to open/close region you can merge multiple buffers/files into one entry. In other words the easiest way so far is to rewrite the archive and skip the entry. But if you find `remove` as a useful/nice to have feature I can start thinking about more efficient implementation. Thanks.

**Comment by Sasszem** [comment]
I think now there's no way of updating an entry in a zip file too. This is an important feature missing.

**Comment by kuba--** [comment]
Features to implement:
- Delete an entry from a zip archive, e.g.:
`int zip_entry_delete(zip, entryname)` and/or `int zip_entry_deletebyindex(zip, index)`

- Update an entry in a zip archive, e.g.:
We may introduce _update_ mode: `zip_open("foo.zip", 'u')` and/or replace current _append_ by generic one _update_. And then you can _seek_ in an entry to overwrite or append data:
```c
zip = zip_open("foo.zip", 'u');

zip_entry_open(zip, "entry.txt");
// the offset is relative to the start of the file, the current position indicator, or end-of-file.
zip_entry_seek(zip, offset, SEEK_END);
zip_entry_write(zip, buf, bufsize);
zip_entry_close(zip);
zip_close(zip);
```

**Comment by Sasszem** [comment]
I think the `zip_entry_open` function should take a third parameter - `mode`.

**Comment by kuba--** [comment]
@Sasszem why `mode` in `zip_entry_open` ?  I think `mode` works globally on _zip archive_ level.
It will require to open _zip archive_ for existing files in 'r' mode by default. You will also handle situations like - someone created a new arhive and tries to open an entry in 'r' mode.

**Comment by Sasszem** [comment]
But that would solve the cases when you want to change an entry, append to an entry and so on.

**Comment by kuba--** [comment]
@Sasszem - yep so what's wrong if you have `zip_open(filename, 'u')` (on the top level), instead of for each entry?
It would be simpler IMO.

**Comment by Sasszem** [comment]
You may want to update a single file, creare a new and delete another, then append to a fourth

**Comment by kuba--** [comment]
So this is exectly what `zip_open(filename, 'u')` may let you do. For instance, right now if you open in append mode (`zip_open(filename, 'a')`) you'll be able to create a new file in the same archive (append) and read existing archives (it's like a reader extended by writer). In other words, it would be similar + delete  and overwrite in place. Actually we'll only need:
* write - which let you overwrite existing archive.
* read - just for extracting (readonly)
* update - combination of both, but if entry exists just overwrite it in place.

**Comment by Sasszem** [comment]
Also, the `'u'`mode is not a standard one

**Comment by Sasszem** [comment]
Oh now I see. That would also do. I would just like  my aplroach better.

**Comment by kuba--** [comment]
Yep, `u` is not a standard, but if you check other zip libraries, it's hard to say they follow any standard.
But what is common is _open mode_. For most of them you set up `mode` on global `zip_archive` level (not per entry), e.g.:
PHP: http://php.net/manual/pl/ziparchive.open.php
`ZipArchive::open ( string $filename [, int $flags ] )` , where:
```
 flags

    The mode to use to open the archive.

        ZIPARCHIVE::OVERWRITE

        ZIPARCHIVE::CREATE

        ZIPARCHIVE::EXCL

        ZIPARCHIVE::CHECKCONS
```

For .NET/C++: https://msdn.microsoft.com/en-us/library
/system.io.compression.ziparchivemode(v=vs.110).aspx

```
ZipArchive(
	Stream stream,
	ZipArchiveMode mode
)

Create:  Only creating new archive entries is permitted.
Read: Only reading archive entries is permitted.
Update: Both read and write operations are permitted for archive entries.
``` 

In Go you just open reader or writer.

To recap. I would keep our API as simple as possible and even simpler ;)
and leave on read and write for `zip_open` (so even get rid of append `a`), but write will work in such a way that if archive exists and you opened an entry which already exists and start writing you'll overwrite data in place (unless you `seek` - move offset to the end of entry, then you'll append to the existing entry. if entry does not exists writing will append a new entry to the archive.

Btw. if you check msdn docs. it says:
`When you set the mode to Update, the underlying file or stream must support reading, writing, and seeking. The content of the entire archive is held in memory, and no data is written to the underlying file or stream until the archive is disposed.`

**Comment by Temtaime** [comment]
Missed entry remove feature forbids me to migrate to this library.
Wanna see it in neat future.

**Comment by kuba--** [comment]
@Temtaime - it's on my list, but I'm rather thinking about replacing `append` by some kind of `update` mode for `zip_open`, what will let append but also remove entries.
Similar to .NET:
```
Create:  Only creating new archive entries is permitted.
Read: Only reading archive entries is permitted.
Update: Both read and write operations are permitted for archive entries.
```
I also didn't find any good backward compatible naming convention, but for `updates` the content of the entire archive will be held in memory, until the archive is closed.

**Comment by Temtaime** [comment]
I think it's bad and only metadata should be stored in memory until it's closed because there can be archives of several gigabytes.

**Comment by kuba--** [comment]
@Temtaime - I understand your concerns (btw. this is exectly what .NET does).
I also do not want to create a temporary copy on disk. Metadata (main directory) is already in memory. Imagine that you remove every second entry from archive. Every time you have to shift data, reindex entries and update main directory. Lets imagine you have 3 entries [0, 1, 2] and you want to remove 2 entries. 
- Could you remove the entry #1 twice? (if yes what would you expect to get)
- Could you remove entries #1 and #2? (because in theory after removing 1st one, the 2nd will get index 1).

I can also mark removed entries but do not change main dictionary until you close the archive. After `close` I can reorder entries.

**Comment by Temtaime** [comment]
Of course, entry #1 can be removed twice. Entry count must be decreased. Is zip structure really needs to shift the data? Maybe we can leave the data as it is with gaps, so the archive will be actually like file was not removed(remove it only from metadata, leaving wasted space). And then we can add function like "repack" to shift data and minimize the size of archive.
Also i don't see any reasons to work with entries. Currently IMO it is only to list archive content. We can drop all those functions working with entry number and make a new one which will call callback passing filename to it for archive content iterating.

**Comment by Nullious** [comment]
ability to delete or overwrite a file inside the archive is a much needed feature.
currently I'm testing writing a settings file into a pre-existing zip for my app, I end up with many copies of the same file inside the zip, so i either need to open the "latest" one, or overwrite the file. obviously delete -> write would be fine too. 
searching for the latest one is a cludge and if you open the zip and delete one of the repeated files ( in zip gui ) they all get deleted as they are the same name eg looking in my zip 
```
"settings.txt"
"settings.txt"
"settings.txt"
"settings.txt"
```

**Comment by amiremohamadi** [comment]
@kuba-- Does anyone works on it?!
I'm using this library as a part of my project and I need to be able to remove entries from zip files.
Any suggestions?

**Comment by kuba--** [comment]
@amiremohamadi - I didn't have time to work on this and, so far. I also didn't hear about anybody who is working on this issue.
So far, the workaround which comes to my mind is to recreate a new archive and rewrite entries. Aftre that delete the original archive.  It's doable without decompressing and compressing again (I hope)

**Comment by sunny-shu** [comment]
@kuba-- 
when use linux api with the mode “O_RDWR”, 
we can lseek the position after the deleting content ,
 then copy the content and lseek  the position to  start of the deleting content  and write the copy content.
So I don't need to create tmp file.I have experimented with it and it is feasible。
i will take a try to solve this problem。

@amiremohamadi 
What is the api you want？

**Comment by kuba--** [comment]
but it will not physically delete any content, it will just remove the entry from global directory

**Comment by jinfeihan57** [comment]
@kuba-- 
The delete function is useful, I want to do this. Do you think if we should support to delete multiple entries at a time (`int zip_entry_delete(zip, entrynamelist)`) or delete only one entry at a time (`int zip_entry_delete(zip, entryname)`)?  The entryname can be a folder name or a file name.

**Comment by kuba--** [comment]
@jinfeihan57 - let's first discuss (design) how we gonna implement it.
The API is a minor thing, right now. It can be one by one or multiple at the same time. Maybe we have to open an entry first to delete it... We can come up with the most useful case later. The main question is - do you want to keep temp. copy in memory? Use the temp file? ...

**Comment by jinfeihan57** [comment]
@kuba-- 
I have two  idea:

1. If I have a file contains 5 entries  [0, 1, 2, 3, 4], and I want to delete #1 and  #3. firstly figure out a new central directory, then move the position of   #2 and  #4 in the file,  so #1 and #3 are overwrited. In the end, put the  new central directory on the file end to cover the old one. 

2.If I have a file contains 5 entries  [0, 1, 2, 3, 4], and I want to delete #1 and  #3.  firstly figure out a new central directory, then copy  #0, #2,  #4 to a temp file and put the  new central directory on the file end, in the end delete the old file and change the temp file name.

**Comment by kuba--** [comment]
@jinfeihan57 - keep in mind that you may `memmove` lot of big entries, e.g. you have N entries:
[1, ..., N][DIR] and you want to remove odd entries 1, 3, 5, ... with different sizes e.g. 1 > 2 > 3 > ... > N. So you will also have to rewrite offsets for all entries and avoid holes/garbage in the archive file.

**Comment by jinfeihan57** [comment]
I implemented entry remove. Anyone interested can try #142 .

**Comment by kuba--** [comment]
Closing this one. `zip_delete_entries` API was added.
